### PR TITLE
Position-angle bug fix that does not introduce new method (based on #5723)

### PR DESCRIPTION
**Pull request by mhvk**
This is a suggested further fix to #5723 (which, annoyingly, I don't seem to be able to make a PR to @dmopalmer's branch for). It does two things:
1. Remove the new `transforms` method, since this is a bug fix and we should not introduce new features.
2. Rename the new `accepting_defaults` keyword argument to `transform_to` to `strict`, which at least to me is more intuitive (@taldcroft?)

This does leave a question of whether it is OK to introduce the new keyword argument. In principle, if anybody subclassed `SkyCoord` and wrote their own `transform_to`, they will now get errors that an extra keyword argument is passed in if they use `separation`.

But I fear this is hard to avoid: we could make a private method `_transform`, but then presumably, if `transform_to` was overridden, it was done so for a reason, so effectively circumventing this would be a bad idea. So, to me, the current approach seems best. @eteq?

### CODE REVIEW

**Comment by mhvk** [comment]
@eteq, @taldcroft - would you have a chance to review this? Something like this is a prerequisite for the nice work by @dmopalmer in introducing his `offset_by` methods (#5727)

**Comment by taldcroft** [comment]
:+1: with this PR in general. 

The main problem I have is in the new kwarg name and docs.  The name `strict` doesn't really convey anything related to the actual meaning of the arg.  It would be nice to think of something that is at least a little specific to what's happening.

The other thing is that the docs don't quite reflect the actual code logic, since the current `strict` actually affects point (1) not point (2).  In words here is the logic:
```
        1. Non-None value in the supplied frame if (``strict`` or not a default)
        2. Non-default value in the SkyCoord instance
        3. Default value in the supplied frame
```
This is crazy confusing and it almost seems necessary to write it out like:
```
If strict:
        1. Non-None value in the supplied frame
        2. Non-default value in the SkyCoord instance

If not strict:
        1. Non-default value in the supplied frame
        2. Non-default value in the SkyCoord instance
        3. Default value in the supplied frame
```
Also, I think it would improve conceptual understanding to use `destination frame` instead of `supplied frame` everywhere.

**Comment by mhvk** [comment]
@taldcroft - as a suggestion, I now went with `override_defaults`, which is closer to @dmopalmer's original `accepting_defaults` but, I think, clearer (the default is `True`, of course). I also extended the docstring, essentially going for your longer version. What do you think?

File: astropy/coordinates/tests/test_sky_coord.py [hunk 1]
```diff
@@ -601,6 +601,16 @@ def test_position_angle():
     assert cicrs.position_angle(cfk5) > 90.0 * u.deg
     assert cicrs.position_angle(cfk5) < 91.0 * u.deg
 
+    # regression check for bug in #5702
+    cfk5B1950 = SkyCoord(1*u.deg, 0*u.deg, frame='fk5', equinox='B1950')
```
Comment (taldcroft) [inline astropy/coordinates/tests/test_sky_coord.py]: Probably need similar test for separation as well?

File: astropy/coordinates/tests/test_sky_coord.py [hunk 2]
```diff
@@ -902,6 +912,16 @@ def test_frame_attr_transform_inherit():
     assert c2.equinox.value == 'J1990.000'
     assert c2.obstime.value == 'J1980.000'
 
+    # The work-around for #5722
+    c  = SkyCoord(1 * u.deg, 2 * u.deg, frame='fk5')
+    c1 = SkyCoord(1 * u.deg, 2 * u.deg, frame='fk5', equinox='B1950.000')
+    c2 = c1.transform_to(c)
+    assert not c2.is_equivalent_frame(c) # counterintuitive, but documented
+    assert c2.equinox.value == 'B1950.000'
+    c3 = c1.transform_to(c, override_defaults=False)
```
Comment (taldcroft) [inline astropy/coordinates/tests/test_sky_coord.py]: Even after dealing with this issue for many days now, I still would get confused when seeing a parameter like `override_defaults=False`.  It just has no meaning to me without looking at the docstring.  But if I saw `prioritize_destination=True`, which is opting in to change the default behavior, I would have a chance to guess/remember what that does.

File: astropy/coordinates/sky_coordinate.py [hunk 3]
```diff
@@ -362,22 +362,32 @@ def _parse_inputs(self, args, kwargs):
...
-        Transform this coordinate to a new frame.
+    def transform_to(self, frame, override_defaults=True):
+        """Transform this coordinate to a new frame.
+
+        The frame attributes (e.g., equinox or obstime) for the returned object
+        depend on the corresponding attributes of both the destination ``frame``
+        and the `SkyCoord` instance being transformed, with a precedence that
+        depends on whether ``override_defaults`` is set or not.
+
+        If ``override_defaults`` is `True`:
```
Comment (taldcroft) [inline astropy/coordinates/sky_coordinate.py]: See comment below about the keyword arg name, but this line should say that this is the default.  But this gets you to the confusing line of
```
If ``override_defaults`` is ``True`` (which is the default):
```
BTW, shouldn't the single backticks around `True` be double ticks?  Or is that actually a sphinx ref?

Comment (bsipocz) [inline astropy/coordinates/sky_coordinate.py]: if single, then it links to https://docs.python.org/3/library/constants.html#True

**Comment by mhvk** [comment]
@taldcroft - so far none of the suggestions for the name have been truly inspiring... It is surely partially because it is already a complicated 3-way process.  Anyway, making a list:
* `accepting_defaults` (bool, `False`) [which defaults?]
* `strict` (bool, `False`) [strict what? my intent was transformation strictly to `frame`]
* `override_defaults` (bool, `True`) [which defaults?]
* `prioritize_destination` (bool, `False`) [prioritize for what?]
* `take_attrs_from` (str, 'source') [but we do take from both]
* `attribute_priority` (str, 'source') [but explicitly set arguments from frame always win]

So, maybe a few new ones:
* `override_frame_defaults` (bool, `True`) [but we don't override all]
* `use_frame_defaults` (bool, `False`) [but we do allow them already]

I must admit that I'm coming back to `strict` as my preferred solution, as to me it feels like telling the `transform_to` method to `strict`ly adhere to the frame I have passed in.

**Comment by taldcroft** [comment]
I think the most typical use of `strict` is to distinguish being strict or liberal in adhering to rules or a standard.  For instance in parsing a VOtable or HTML, one can be `strict` and fail on inputs that do not adhere to the defined standard, or let those pass and take a good guess.  Likewise `Unit` has `parse_strict` which serves roughly the same purpose.  That kind of usage would not apply here because the new keyword is being used to *change* the rules, not whether to *enforce* them.

So `prioritize_dest_attrs`?   Or since it usually occurs in otherwise short code lines, `prioritize_destintation_attrs`?  (`attrs` is a standard abbreviation in coordinates).

**Comment by mhvk** [comment]
I ended up back at `strict` -- I think in the end this does not matter that much, and it will be used rarely enough that if one sees it in code, one has to look up the docstring anyway...  But I do think if it gives the impression "strictly follow the rules", it does what it is supposed to: strictly transform to `frame`, no iffing and butting.

**Comment by mhvk** [comment]
@taldcroft - to avoid `strict`, how about `exact`? This would be in the sense of `use_frame_exactly_as_is` (and thus might also be `exactly_as_is` or `use_as_is` or `as_is`). I'm still trying to convey with the name what the effect is rather than how things are implemented. (Note that now that "extra" parameters are stored on the `SkyCoord` rather than its frame, this is in fact what the argument enforces: for the non-default case, in the resulting frame, no attributes arise from the source.)

@eteq - you earlier have several times pointed to much nicer ways of working arguments; for the particular case of `SkyCoord.transform_to` here, where we need a keyword to say "please don't use the source attributes to override default frame attributes", @taldcroft and I seem to have been unable to come up with anything that pleases both of us (for a list of our trials, see https://github.com/astropy/astropy/pull/5762#issuecomment-278758618).

**Comment by taldcroft** [comment]
@mhvk - I think the problem with all of your short name suggestions is that it assumes someone reading the code has the context that we now have, namely that the arg refers to treatment of frame attributes.  

But imagine otherwise and seeing `c_fk4 = c.transform_to('fk4', exact=False)`.  My reaction would be "What?!?  Is it using some approximate transform?".  If I were too lazy to go read the docs that might stick in my head enough to later write `c_fk4 = c.transform_to('fk4', exact=True)` to ensure that no approximations are used.  So I really think we should go with some long kwarg that cannot easily be misinterpreted, even if its true meaning still requires reading the docs.

**Comment by dmopalmer** [comment]
I agree that we don't want a terse, uninformative parameter name that usually means something else (especially something that is relevant in this context).

With modern IDEs, you could have call the parameter `taking_every_single_aspect_of_the_destination_frame_as_being_identical_to_the_model_frame=`, and people would still use it: `tak<tab>`.

Also, due to the magic of the [anchoring effect](https://en.wikipedia.org/wiki/Anchoring) you will now say "89 characters is too long for a parameter name, let's see if we can name it in 40 characters".

**Comment by mhvk** [comment]
OK, if we go for a long name, how about my above suggestions of `use_frame_exactly_as_is` (or a little shorter, `use_frame_as_is`, `use_as_is`, or even `as_is`).

**Comment by taldcroft** [comment]
I still would like to emphasize that the new arg is specifying a *merge policy for two sets of attributes*.  "Use frame exactly" is not what is happening, first off because there are two frames,  and second `use_frame_exactly=False` would imply using the frame inexactly, which again raises alarms in my head.

`attributes_merge_strategy=('source' (default) | 'destination')` ?

This has some heritage with [metadata merging](http://docs.astropy.org/en/stable/api/astropy.utils.metadata.enable_merge_strategies.html#astropy.utils.metadata.enable_merge_strategies).

**Comment by mhvk** [comment]
This probably is where our difference arises: there is not really a merge strategy in the non-default case: the destination frame will have *no* attributes from the source frame, i.e., you are guaranteed that your coordinates are in exactly the frame passed in (this is a consequence of my PR that made sure attributes irrelevant to a frame are no longer stored on the frame).

I also note it is this need for transforing to exactly the given frame is what led to the PR in the first place: it is essential for getting separations and position angles correct. My sense is thus that it is also this need that should be conveyed by the name of the argument (which is why I may seem so stuck on "exact", "strict" etc.)

**Comment by taldcroft** [comment]
> the destination frame will have *no* attributes from the source frame

@mhvk - Hopefully I'm not making a mistake reading the code (not in a place to just try right now), but if you transform from FK4 with a non-default `obstime` to ICRS then I think you will still have `obstime` on the end result (even though the container on which it is stored may be different now).  That's the intended behavior of SkyCoord attribute handling, namely moving non-frame attributes around to the extent possible according to some rules.  We are just adding a variation of those rules.

As a matter of semantics, even choosing *only* attributes from the "left" side is still a perfectly valid merge strategy, and the default is definitely a merge.

**Comment by mhvk** [comment]
@taldcroft - we should get this in for 1.3.1. In what I hope is a final change, I now made it `merge_attributes=<bool>` (default `True`), which does reflect your idea of merging, but keeps my sense of it being boolena, and that `False` means the destination *frame* gets none of the source's attributes. I hope this is an OK compromise, and that the logic is clear from the rewritten docstring:
```
    def transform_to(self, frame, merge_attributes=True):
        """Transform this coordinate to a new frame.

        The precise frame transformed to depends on ``merge_attributes``.
        If `False`, the destination frame is used exactly as passed in.
        But this is often not quite what one wants.  E.g., suppose one wants to
        transform an ICRS coordinate that has an obstime attribute to FK4; in
        this case, one likely would want to use this information. Thus, the
        default for ``merge_attributes`` is `True`, in which the precedence is
        as follows: (1) explicitly set (i.e., non-default) values in the
        destination frame; (2) explicitly set values in the source; (3) default
        value in the destination frame.

        Note that in either case, any explicitly set attributes on the source
        `SkyCoord` that are not part of the destination frame's definition are
        kept (stored on the resulting `SkyCoord`), and thus one can round-trip
        (e.g., from FK4 to ICRS to FK4 without loosing obstime).
```

**Comment by taldcroft** [comment]
@mhvk - excellent, I'm happy with your final suggestion of `merge_attributes`.

File: astropy/coordinates/sky_coordinate.py [hunk 4]
```diff
@@ -900,9 +903,13 @@ def match_to_catalog_3d(self, catalogcoord, nthneighbor=1):
...
         from .matching import match_coordinates_3d
 
+
         if (isinstance(catalogcoord, (SkyCoord, BaseCoordinateFrame))
                 and catalogcoord.has_data):
-            self_in_catalog_frame = self.transform_to(catalogcoord)
+            if self.is_equivalent_frame(catalogcoord):
+                self_in_catalog_frame = self
+            else:
+                self_in_catalog_frame = self.transform_to(catalogcoord)
```
Comment (taldcroft) [inline astropy/coordinates/sky_coordinate.py]: I haven't thought about this carefully, but should this have `merge_attributes=False`?

File: astropy/coordinates/tests/test_sky_coord.py (hunk 1 above)
Comment (taldcroft) [inline astropy/coordinates/tests/test_sky_coord.py]: Still think that each of the functions that was fixed needs a test (e.g. using the backwards/forwards equivalence method using for position angle).

File: astropy/coordinates/sky_coordinate.py (hunk 4 above)
Comment (mhvk) [inline astropy/coordinates/sky_coordinate.py]: Hmm, yes, and then presumably also for `match_to_catalog_sky` (which was not changed in #5723)

Comment (mhvk) [inline astropy/coordinates/sky_coordinate.py]: Now, looking at `matching,py`, I see that the first think `match_coordinates_sky` does is to transform the coordinate to the catalog frame: https://github.com/astropy/astropy/blob/master/astropy/coordinates/matching.py#L72
So, these lines should not be here in the first place and we should adjust the code in `matching.py`.

I don't really have time to delve into this now; shall I just undo this change and raise a new issue?

File: astropy/coordinates/tests/test_sky_coord.py (hunk 1 above)
Comment (mhvk) [inline astropy/coordinates/tests/test_sky_coord.py]: OK, will do.

File: astropy/coordinates/sky_coordinate.py (hunk 4 above)
Comment (taldcroft) [inline astropy/coordinates/sky_coordinate.py]: > I don't really have time to delve into this now; shall I just undo this change and raise a new issue?

Fine by me.

Comment (mhvk) [inline astropy/coordinates/sky_coordinate.py]: I made a new commit that reverts the change and opened #5857 to remind us that `match_*` needs some work.

**Comment by mhvk** [comment]
Hopefully all OK now (but ensure checks on one-but-last commit pass)

**Comment by mhvk** [comment]
The one travis failure was (yet another) URL timeout. I restarted just in case, but really all should be OK.
